src/main.js

```diff
@@ -150,39 +153,97 @@ $(document).ready(() => {
       errorView.show(err);
     }
 
-    function toggleSidebarAndSave() {
-      store.set(STORE.SHOWN, !isSidebarVisible(), () => {
-        toggleSidebar();
-        if (isSidebarVisible()) {
-          tryLoadRepo();
-        }
-      });
-    }
-
     function toggleSidebar(visibility) {
       if (visibility !== undefined) {
         if (isSidebarVisible() === visibility) return;
         toggleSidebar();
       } else {
         $html.toggleClass(SHOW_CLASS);
         $document.trigger(EVENT.TOGGLE, isSidebarVisible());
+        // Load repo when the page first loaded in the layout mode
+        if (isSidebarVisible()) tryLoadRepo();
+      }
+
+      return visibility;
+    }
+
+    function togglePin(isPinned) {
+      if (isPinned !== undefined) {
+        if (isSidebarPinned() === isPinned) return;
+        return togglePin();
       }
+
+      $pinner.toggleClass(PINNED_CLASS);
+      const sidebarPinned = isSidebarPinned();
+      $pinner.find('.tooltipped').attr('aria-label', `${sidebarPinned ? 'Pin' : 'Unpin'} octotree to the page`);
+
+      $document.trigger(EVENT.PIN, sidebarPinned);
+
+      store.set(STORE.PINNED, sidebarPinned);
+
+      if (sidebarPinned) toggleSidebar(true);
+
+      return sidebarPinned;
     }
 
     function layoutChanged(save = false) {
       const width = $sidebar.outerWidth();
-      adapter.updateLayout(isTogglerVisible(), isSidebarVisible(), width);
+      adapter.updateLayout(isSidebarPinned(), isSidebarVisible(), width);
       if (save === true) {
         store.set(STORE.WIDTH, width);
       }
     }
 
+    /**
+     * Handling auto toggle behaviours
+     *  - The toggle timer starts over whenever user leave/navigate the sidebar
+     *  - Clearing the toggle timer on mousemove handles the case
+     *    in which mouse moves out and in the sidebard during the delay time
+     *  - There is no delay when using shortkeys
+     */
+    function handleSidebarAutoToggling() {
+      let timerId = null;
+
+      const resetTimer = (delay) => {
+        if (!isSidebarPinned()) {
+          clearTimer();
+          // Using isSidebarPinned as a flag to ensure the UI consistency
+          timerId = setTimeout(() => toggleSidebar(isSidebarPinned()), delay);
+        }
+      };
+
+      const clearTimer = () => timerId && clearTimeout(timerId);
+
+      $sidebar.on('keyup mouseleave', () => resetTimer(SIDEBAR_HIDING_DELAY));
```

Comment:
`keyup` will trigger the timer so that the sidebar will be closed when users stop navigating the sidebar within the delay. I think that is an expected behaviour. If `keyup` moves to the line below, the sidebar will never be closed after user navigates the sidebar